src/plonk/prover.rs

```diff
@@ -20,256 +20,399 @@ use crate::transcript::TranscriptWrite;
-    circuit.synthesize(&mut witness, config)?;
-
-    let witness = witness;
-
-    // Compute commitments to aux column polynomials
-    let aux_commitments_projective: Vec<_> = aux
-        .iter()
-        .map(|poly| params.commit_lagrange(poly, Blind::default()))
-        .collect();
-    let mut aux_commitments = vec![C::zero(); aux_commitments_projective.len()];
-    C::Projective::batch_to_affine(&aux_commitments_projective, &mut aux_commitments);
-    let aux_commitments = aux_commitments;
-    drop(aux_commitments_projective);
-    metrics::counter!("aux_commitments", aux_commitments.len() as u64);
-
-    for commitment in &aux_commitments {
-        transcript
-            .common_point(*commitment)
-            .map_err(|_| Error::TranscriptError)?;
+    struct AuxSingle<C: CurveAffine> {
+        pub aux_values: Vec<Polynomial<C::Scalar, LagrangeCoeff>>,
+        pub aux_polys: Vec<Polynomial<C::Scalar, Coeff>>,
+        pub aux_cosets: Vec<Polynomial<C::Scalar, ExtendedLagrangeCoeff>>,
     }
 
-    let aux_polys: Vec<_> = aux
+    let aux_vec: Result<Vec<_>, _> = auxs
         .iter()
-        .map(|poly| {
-            let lagrange_vec = domain.lagrange_from_vec(poly.to_vec());
-            domain.lagrange_to_coeff(lagrange_vec)
-        })
-        .collect();
+        .map(|aux| -> Result<AuxSingle<C>, Error> {
+            let aux_commitments_projective: Vec<_> = aux
+                .iter()
+                .map(|poly| params.commit_lagrange(poly, Blind::default()))
+                .collect();
+            let mut aux_commitments = vec![C::zero(); aux_commitments_projective.len()];
+            C::Projective::batch_to_affine(&aux_commitments_projective, &mut aux_commitments);
+            let aux_commitments = aux_commitments;
+            drop(aux_commitments_projective);
+            metrics::counter!("aux_commitments", aux_commitments.len() as u64);
+
+            for commitment in &aux_commitments {
+                transcript
+                    .common_point(*commitment)
+                    .map_err(|_| Error::TranscriptError)?;
+            }
+
+            let aux_polys: Vec<_> = aux
+                .iter()
+                .map(|poly| {
+                    let lagrange_vec = domain.lagrange_from_vec(poly.to_vec());
+                    domain.lagrange_to_coeff(lagrange_vec)
+                })
+                .collect();
 
-    let aux_cosets: Vec<_> = meta
-        .aux_queries
-        .iter()
-        .map(|&(column, at)| {
-            let poly = aux_polys[column.index()].clone();
-            domain.coeff_to_extended(poly, at)
+            let aux_cosets: Vec<_> = meta
+                .aux_queries
+                .iter()
+                .map(|&(column, at)| {
+                    let poly = aux_polys[column.index()].clone();
+                    domain.coeff_to_extended(poly, at)
+                })
+                .collect();
+
+            Ok(AuxSingle {
+                aux_values: aux.to_vec(),
```

Comment:
```suggestion
                aux_values: *aux,
```

This avoids the allocation done by `to_vec()`